src/parquet/arrow/reader.cc

```diff
@@ -152,6 +153,64 @@ class SingleRowGroupIterator : public FileColumnIterator {
   bool done_;
 };
 
+class RowGroupRecordBatchReader : public ::arrow::RecordBatchReader {
+ public:
+    explicit RowGroupRecordBatchReader(const std::vector<int>& row_group_indices,
+                                       const std::vector<int>& column_indices,
+                                       FileReader* reader)
+      : row_group_indices_(row_group_indices),
+        column_indices_(column_indices),
+        file_reader_(reader),
+        next_row_group_(0) {
+      file_reader_->GetSchema(column_indices_, &schema_);
+    }
+
+    ~RowGroupRecordBatchReader() {}
+
+    std::shared_ptr<::arrow::Schema> schema() const override {
+      return schema_;
+    }
+
+    Status ReadNext(std::shared_ptr<::arrow::RecordBatch> *out) override {
+      if (table_ != nullptr) { // one row group has been loaded
+        std::shared_ptr<::arrow::RecordBatch> tmp;
+        table_batch_reader_->ReadNext(&tmp);
+        if (tmp != nullptr) { // some column chunks are left in table
+          *out = tmp;
+          return Status::OK();
+        } else { // the entire table is consumed
+          table_batch_reader_.reset();
+          table_.reset();
+        }
+      }
+
+      // all row groups has been consumed
+      if (next_row_group_ == row_group_indices_.size()) {
+        *out = nullptr;
+        return Status::OK();
+      }
+
+      RETURN_NOT_OK(file_reader_->ReadRowGroup(row_group_indices_[next_row_group_],
```

Comment:
I am most concern about this one. We have to read one entire row group, but the caller may consume only the first N RecordBatches.

I am wondering that this is not optimal